File: src/services/summary_service.py
"""
Monthly consumption summary service.
Aggregates purchased quantities from intake records and merges with
consumed quantities entered by managers to calculate waste.
"""
from src.services.intake_service import _INTAKE_RECORDS

# In-memory consumption store: { (restaurant_id, item_id, month): units_consumed }
_CONSUMPTION: dict[tuple, int] = {
    ("R001", "SKU001", "2026-03"): 160,
    ("R001", "SKU002", "2026-03"): 190,
    ("R001", "SKU003", "2026-03"): 60,
    ("R001", "SKU004", "2026-02"): 35,
}

WASTE_THRESHOLDS = {"green": 15, "amber": 30}  # % boundaries


def _waste_level(waste_pct: float) -> str:
    if waste_pct <= WASTE_THRESHOLDS["green"]:
        return "green"
    if waste_pct <= WASTE_THRESHOLDS["amber"]:
        return "amber"
    return "red"


def get_monthly_summary(restaurant_id: str, month: str) -> dict:
    """
    Aggregate all purchased units for the month, merge with consumed
    quantities, and compute waste per item.
    """
    # Aggregate purchased per item_id for the month
    purchased: dict[str, dict] = {}
    for r in _INTAKE_RECORDS:
        if r["restaurant_id"] == restaurant_id and r["delivery_date"].startswith(month):
            key = r["item_id"]
            if key not in purchased:
                purchased[key] = {
                    "item_id": key,
                    "name": r["name"],
                    "category": r["category"],
                    "units_purchased": 0,
                }
            purchased[key]["units_purchased"] += r["units"]

    # Build summary rows
    items = []
    for item_id, p in sorted(purchased.items()):
        consumed = _CONSUMPTION.get((restaurant_id, item_id, month), 0)
        waste = max(p["units_purchased"] - consumed, 0)
        waste_pct = round((waste / p["units_purchased"]) * 100, 1) if p["units_purchased"] else 0.0
        items.append({
            "item_id": item_id,
            "name": p["name"],
            "category": p["category"],
            "units_purchased": p["units_purchased"],
            "units_consumed": consumed,
            "waste": waste,
            "waste_pct": waste_pct,
            "waste_level": _waste_level(waste_pct),
        })

    total_purchased = sum(i["units_purchased"] for i in items)
    total_consumed = sum(i["units_consumed"] for i in items)
    total_waste = max(total_purchased - total_consumed, 0)
    overall_waste_pct = round((total_waste / total_purchased) * 100, 1) if total_purchased else 0.0

    return {
        "restaurant_id": restaurant_id,
        "month": month,
        "items": items,
        "total_purchased": total_purchased,
        "total_consumed": total_consumed,
        "total_waste": total_waste,
        "overall_waste_pct": overall_waste_pct,
        "overall_waste_level": _waste_level(overall_waste_pct),
    }


def save_consumption(restaurant_id: str, month: str, entries: list[dict]) -> None:
    """
    Save or update consumed quantities for a month.
    Each entry: { item_id, units_consumed }
    """
    for entry in entries:
        _CONSUMPTION[(restaurant_id, entry["item_id"], month)] = entry["units_consumed"]

File: backend-dotnet/src/InventoryWastePredictor.Api/Services/SummaryService.cs
namespace InventoryWastePredictor.Api.Services;

public class SummaryService
{
    private static readonly int WasteThresholdGreen = 15;
    private static readonly int WasteThresholdAmber = 30;

    internal readonly Dictionary<(string RestaurantId, string ItemId, string Month), int> Consumption = new()
    {
        [("R001", "SKU001", "2026-03")] = 160,
        [("R001", "SKU002", "2026-03")] = 190,
        [("R001", "SKU003", "2026-03")] = 60,
        [("R001", "SKU004", "2026-02")] = 35,
    };

    private readonly IntakeService _intakeService;

    public SummaryService(IntakeService intakeService)
    {
        _intakeService = intakeService;
    }

    private static string WasteLevel(double wastePct) =>
        wastePct <= WasteThresholdGreen ? "green" :
        wastePct <= WasteThresholdAmber ? "amber" : "red";

    public Dictionary<string, object> GetMonthlySummary(string restaurantId, string month)
    {
        // Aggregate purchased per item_id for the month
        var purchased = new Dictionary<string, (string Name, string Category, int Units)>();
        foreach (var r in _intakeService.IntakeRecords)
        {
            if (r.RestaurantId == restaurantId && r.DeliveryDate.StartsWith(month))
            {
                if (!purchased.ContainsKey(r.ItemId))
                    purchased[r.ItemId] = (r.Name, r.Category, 0);
                var entry = purchased[r.ItemId];
                purchased[r.ItemId] = (entry.Name, entry.Category, entry.Units + r.Units);
            }
        }

        var items = new List<Dictionary<string, object>>();
        foreach (var (itemId, p) in purchased.OrderBy(kv => kv.Key))
        {
            var consumed = Consumption.TryGetValue((restaurantId, itemId, month), out var c) ? c : 0;
            var waste = Math.Max(p.Units - consumed, 0);
            var wastePct = p.Units > 0 ? Math.Round((double)waste / p.Units * 100, 1) : 0.0;
            items.Add(new Dictionary<string, object>
            {
                ["item_id"]        = itemId,
                ["name"]           = p.Name,
                ["category"]       = p.Category,
                ["units_purchased"] = p.Units,
                ["units_consumed"] = consumed,
                ["waste"]          = waste,
                ["waste_pct"]      = wastePct,
                ["waste_level"]    = WasteLevel(wastePct),
            });
        }

        var totalPurchased = items.Sum(i => (int)i["units_purchased"]);
        var totalConsumed  = items.Sum(i => (int)i["units_consumed"]);
        var totalWaste     = Math.Max(totalPurchased - totalConsumed, 0);
        var overallWastePct = totalPurchased > 0
            ? Math.Round((double)totalWaste / totalPurchased * 100, 1) : 0.0;

        return new Dictionary<string, object>
        {
            ["restaurant_id"]      = restaurantId,
            ["month"]              = month,
            ["items"]              = items,
            ["total_purchased"]    = totalPurchased,
            ["total_consumed"]     = totalConsumed,
            ["total_waste"]        = totalWaste,
            ["overall_waste_pct"]  = overallWastePct,
            ["overall_waste_level"] = WasteLevel(overallWastePct),
        };
    }

    public void SaveConsumption(string restaurantId, string month,
                                IEnumerable<(string ItemId, int UnitsConsumed)> entries)
    {
        foreach (var (itemId, units) in entries)
            Consumption[(restaurantId, itemId, month)] = units;
    }
}
